File: packages/memory_server/memory_server/api/v1/documents.py
"""Document ingest API."""

from __future__ import annotations

from typing import Annotated, Any

from fastapi import APIRouter, Depends, Header, Query, Response, status
from memory_core.application import (
    DeleteDocumentCommand,
    GetDocumentQuery,
    IngestDocumentCommand,
    ListDocumentChunksQuery,
    ProcessDocumentCommand,
)
from memory_core.domain.entities import MemoryChunk, MemoryDocument
from pydantic import BaseModel, Field

from memory_server.api.auth import require_service_token
from memory_server.api.dependencies import get_container
from memory_server.api.policy import ensure_server_writes_enabled
from memory_server.api.v1.scope_resolution import resolve_single_scope
from memory_server.backpressure import document_ingest_backpressure_response
from memory_server.composition import Container
from memory_server.pagination import cursor_int, cursor_str, decode_cursor, encode_cursor

router = APIRouter(
    prefix="/documents",
    tags=["documents"],
    dependencies=[Depends(require_service_token)],
)


class IngestDocumentRequest(BaseModel):
    space_id: str | None = Field(default=None, min_length=1, max_length=80)
    profile_id: str | None = Field(default=None, min_length=1, max_length=80)
    thread_id: str | None = Field(default=None, max_length=80)
    space_slug: str | None = Field(default=None, min_length=1, max_length=160)
    profile_external_ref: str | None = Field(default=None, min_length=1, max_length=200)
    thread_external_ref: str | None = Field(default=None, min_length=1, max_length=200)
    title: str = Field(min_length=1, max_length=300)
    text: str = Field(min_length=1, max_length=500_000)
    source_type: str = Field(default="document", min_length=1, max_length=80)
    source_external_id: str = Field(min_length=1, max_length=240)
    classification: str = Field(default="unknown", max_length=40)


def document_to_response(
    document: MemoryDocument,
    *,
    chunks: int | None = None,
    duplicate_chunks: int | None = None,
    indexing_status: str | None = None,
    deleted_chunks: int | None = None,
    deleted_facts: int | None = None,
) -> dict[str, Any]:
    body: dict[str, Any] = {
        "id": str(document.id),
        "space_id": str(document.space_id),
        "profile_id": str(document.profile_id),
        "thread_id": str(document.thread_id) if document.thread_id else None,
        "title": document.title,
        "source_type": document.source_type,
        "source_external_id": document.source_external_id,
        "content_hash": document.content_hash,
        "classification": document.classification,
        "status": document.status.value,
        "created_at": document.created_at.isoformat(),
        "updated_at": document.updated_at.isoformat(),
    }
    if chunks is not None:
        body["chunks"] = chunks
    if duplicate_chunks is not None:
        body["duplicate_chunks"] = duplicate_chunks
    if indexing_status is not None:
        body["indexing_status"] = indexing_status
    if deleted_chunks is not None:
        body["deleted_chunks"] = deleted_chunks
    if deleted_facts is not None:
        body["deleted_facts"] = deleted_facts
    return body


def chunk_to_response(chunk: MemoryChunk) -> dict[str, Any]:
    return {
        "id": str(chunk.id),
        "document_id": str(chunk.document_id) if chunk.document_id else None,
        "text": chunk.text,
        "kind": chunk.kind.value,
        "sequence": chunk.sequence,
        "status": chunk.status.value,
        "classification": chunk.classification,
    }


@router.post("", status_code=status.HTTP_201_CREATED)
async def ingest_document(
    request: IngestDocumentRequest,
    container: Annotated[Container, Depends(get_container)],
    response: Response,
    idempotency_key: Annotated[str | None, Header(alias="Idempotency-Key")] = None,
) -> Any:
    ensure_server_writes_enabled(container)
    backpressure = await document_ingest_backpressure_response(container)
    if backpressure is not None:
        return backpressure
    scope = await resolve_single_scope(
        container,
        space_id=request.space_id,
        profile_id=request.profile_id,
        thread_id=request.thread_id,
        space_slug=request.space_slug,
        profile_external_ref=request.profile_external_ref,
        thread_external_ref=request.thread_external_ref,
        thread_required=False,
    )
    result = await container.ingest_document.execute(
        IngestDocumentCommand(
            space_id=scope.space_id,
            profile_id=scope.profile_id,
            thread_id=scope.thread_id,
            title=request.title,
            text=request.text,
            source_type=request.source_type,
            source_external_id=request.source_external_id,
            idempotency_key=idempotency_key,
            classification=request.classification,
        )
    )
    if result.indexing_status == "already_indexed_or_pending":
        response.status_code = status.HTTP_200_OK
    return {
        "data": document_to_response(
            result.document,
            chunks=len(result.chunks),
            duplicate_chunks=result.duplicate_chunks,
            indexing_status=result.indexing_status,
        )
    }


@router.get("/{document_id}")
async def get_document(
    document_id: str,
    container: Annotated[Container, Depends(get_container)],
) -> dict[str, Any]:
    result = await container.get_document.execute(GetDocumentQuery(document_id=document_id))
    return {"data": document_to_response(result.document)}


@router.get("/{document_id}/chunks")
async def list_document_chunks(
    document_id: str,
    container: Annotated[Container, Depends(get_container)],
    limit: Annotated[int, Query(ge=1, le=500)] = 100,
    cursor: Annotated[str | None, Query(max_length=1000)] = None,
) -> dict[str, Any]:
    decoded_cursor = decode_cursor(cursor, kind="document_chunks")
    result = await container.list_document_chunks.execute(
        ListDocumentChunksQuery(
            document_id,
            limit=limit + 1,
            cursor_sequence=cursor_int(decoded_cursor, "sequence"),
            cursor_id=cursor_str(decoded_cursor, "id"),
        )
    )
    chunks = list(result.chunks)
    visible_chunks = chunks[:limit]
    next_cursor = None
    if len(chunks) > limit and visible_chunks:
        last = visible_chunks[-1]
        next_cursor = encode_cursor(
            "document_chunks",
            sequence=last.sequence,
            id=str(last.id),
        )
    return {
        "data": [chunk_to_response(chunk) for chunk in visible_chunks],
        "next_cursor": next_cursor,
    }


@router.post("/{document_id}/process")
async def process_document(
    document_id: str,
    container: Annotated[Container, Depends(get_container)],
    idempotency_key: Annotated[str | None, Header(alias="Idempotency-Key")] = None,
) -> dict[str, Any]:
    ensure_server_writes_enabled(container)
    result = await container.process_document.execute(
        ProcessDocumentCommand(document_id=document_id, idempotency_key=idempotency_key)
    )
    return {
        "data": document_to_response(
            result.document,
            chunks=result.chunks,
            indexing_status=result.indexing_status,
        )
    }


@router.delete("/{document_id}")
async def delete_document(
    document_id: str,
    container: Annotated[Container, Depends(get_container)],
) -> dict[str, Any]:
    ensure_server_writes_enabled(container)
    result = await container.delete_document.execute(DeleteDocumentCommand(document_id=document_id))
    return {
        "data": document_to_response(
            result.document,
            deleted_chunks=result.deleted_chunks,
            deleted_facts=result.deleted_facts,
            indexing_status=result.indexing_status,
        )
    }

File: packages/infinity_context_ts_sdk/src/resources/documents.ts
import type { RequestExecutor } from "../client.js";
import { scopeQuery, singleScopePayload, withoutUndefined, type SingleScopeInput } from "../payload.js";
import type { ApiEnvelope, DocumentRecord, JsonObject, SourceRef } from "../types.js";

export class DocumentsClient {
  constructor(private readonly http: RequestExecutor) {}

  ingestDocument(input: SingleScopeInput & {
    readonly title: string;
    readonly text: string;
    readonly sourceExternalId: string;
    readonly sourceType?: string;
    readonly classification?: string;
    readonly sourceRefs?: readonly SourceRef[];
    readonly idempotencyKey?: string;
  }): Promise<ApiEnvelope<DocumentRecord>> {
    return this.http.request<ApiEnvelope<DocumentRecord>>({
      method: "POST",
      path: "/v1/documents",
      idempotencyKey: input.idempotencyKey,
      json: withoutUndefined({
        ...singleScopePayload(input),
        title: input.title,
        text: input.text,
        source_type: input.sourceType ?? "document",
        source_external_id: input.sourceExternalId,
        classification: input.classification ?? "unknown",
        source_refs: input.sourceRefs,
      }) as JsonObject,
    });
  }

  ingestEpisode(input: SingleScopeInput & {
    readonly sourceExternalId: string;
    readonly text: string;
    readonly sourceType?: string;
    readonly occurredAt?: string;
    readonly speaker?: string;
    readonly trustLevel?: string;
    readonly kindHint?: string;
    readonly language?: string;
    readonly metadata?: JsonObject;
    readonly idempotencyKey?: string;
  }): Promise<ApiEnvelope<JsonObject>> {
    return this.http.request<ApiEnvelope<JsonObject>>({
      method: "POST",
      path: "/v1/episodes",
      idempotencyKey: input.idempotencyKey,
      json: withoutUndefined({
        ...singleScopePayload(input),
        source_type: input.sourceType ?? "unknown",
        source_external_id: input.sourceExternalId,
        text: input.text,
        occurred_at: input.occurredAt,
        speaker: input.speaker,
        trust_level: input.trustLevel ?? "medium",
        kind_hint: input.kindHint,
        language: input.language,
        metadata: input.metadata,
      }) as JsonObject,
    });
  }

  getDocument(documentId: string): Promise<ApiEnvelope<DocumentRecord>> {
    return this.http.request<ApiEnvelope<DocumentRecord>>({
      method: "GET",
      path: `/v1/documents/${documentId}`,
    });
  }

  listDocumentChunks(
    documentId: string,
    input: { readonly limit?: number; readonly cursor?: string } = {},
  ): Promise<ApiEnvelope<JsonObject[]>> {
    return this.http.request<ApiEnvelope<JsonObject[]>>({
      method: "GET",
      path: `/v1/documents/${documentId}/chunks`,
      params: withoutUndefined({ limit: input.limit ?? 100, cursor: input.cursor }),
    });
  }

  processDocument(
    documentId: string,
    input: { readonly idempotencyKey?: string } = {},
  ): Promise<ApiEnvelope<DocumentRecord>> {
    return this.http.request<ApiEnvelope<DocumentRecord>>({
      method: "POST",
      path: `/v1/documents/${documentId}/process`,
      idempotencyKey: input.idempotencyKey,
    });
  }

  deleteDocument(documentId: string): Promise<ApiEnvelope<DocumentRecord>> {
    return this.http.request<ApiEnvelope<DocumentRecord>>({
      method: "DELETE",
      path: `/v1/documents/${documentId}`,
    });
  }

  listScopeDocuments(input: SingleScopeInput & { readonly limit?: number }): Promise<ApiEnvelope<DocumentRecord[]>> {
    return this.http.request<ApiEnvelope<DocumentRecord[]>>({
      method: "GET",
      path: "/v1/documents",
      params: withoutUndefined({ ...scopeQuery(input), limit: input.limit ?? 100 }),
    });
  }
}
